agent/proxycfg/api_gateway.go

```diff
@@ -0,0 +1,492 @@
+	if u.Err != nil {
+		return fmt.Errorf("error filling agent cache: %v", u.Err)
+	}
+
+	switch {
+	case u.CorrelationID == rootsWatchID:
+		// Handle change in the CA roots
+		if err := h.handleRootCAUpdate(u, snap); err != nil {
+			return err
+		}
+	case u.CorrelationID == gatewayConfigWatchID:
+		// Handle change in the api-gateway or bound-api-gateway config entry
+		if err := h.handleGatewayConfigUpdate(ctx, u, snap); err != nil {
+			return err
+		}
+	case u.CorrelationID == inlineCertificateConfigWatchID:
+		// Handle change in an attached inline-certificate config entry
+		if err := h.handleInlineCertConfigUpdate(ctx, u, snap); err != nil {
+			return err
+		}
+	case u.CorrelationID == routeConfigWatchID:
+		// Handle change in an attached http-route or tcp-route config entry
+		if err := h.handleRouteConfigUpdate(ctx, u, snap); err != nil {
+			return err
+		}
+	default:
+		return (*handlerUpstreams)(h).handleUpdateUpstreams(ctx, u, snap)
+	}
+
+	return nil
+}
+
+// handleRootCAUpdate responds to changes in the watched root CA for a gateway
+func (h *handlerAPIGateway) handleRootCAUpdate(u UpdateEvent, snap *ConfigSnapshot) error {
+	roots, ok := u.Result.(*structs.IndexedCARoots)
+	if !ok {
+		return fmt.Errorf("invalid type for response: %T", u.Result)
+	}
+	snap.Roots = roots
+	return nil
+}
+
+// handleGatewayConfigUpdate responds to changes in the watched config entry for a gateway.
+// In particular, we want to make sure that we're subscribing to any attached resources such
+// as routes and certificates. These additional subscriptions will enable us to update the
+// config snapshot appropriately for any route or certificate changes.
+func (h *handlerAPIGateway) handleGatewayConfigUpdate(ctx context.Context, u UpdateEvent, snap *ConfigSnapshot) error {
+	resp, ok := u.Result.(*structs.ConfigEntryResponse)
+	if !ok {
+		return fmt.Errorf("invalid type for response: %T", u.Result)
+	} else if resp.Entry == nil {
+		return nil
+	}
+
+	switch gwConf := resp.Entry.(type) {
+	case *structs.BoundAPIGatewayConfigEntry:
+		snap.APIGateway.BoundGatewayConfig = gwConf
+
+		seenRefs := make(map[structs.ResourceReference]any)
+		for _, listener := range gwConf.Listeners {
+			snap.APIGateway.BoundListeners[listener.Name] = listener
+
+			// Subscribe to changes in each attached x-route config entry
+			for _, ref := range listener.Routes {
+				seenRefs[ref] = struct{}{}
+
+				ctx, cancel := context.WithCancel(ctx)
+				switch ref.Kind {
+				case structs.HTTPRoute:
+					snap.APIGateway.HTTPRoutes.InitWatch(ref, cancel)
+				case structs.TCPRoute:
+					snap.APIGateway.TCPRoutes.InitWatch(ref, cancel)
+				default:
+					cancel()
+					return fmt.Errorf("unexpected route kind on gateway: %s", ref.Kind)
+				}
+
+				err := h.subscribeToConfigEntry(ctx, ref.Kind, ref.Name, routeConfigWatchID)
+				if err != nil {
+					// TODO May want to continue
+					return err
```

Comment:
Just wondering for familiarity sake, what happens when we error here? Do we just retry the event handler execution?

Reply:
Looks like we just log and wait for the next watch trigger [here](https://github.com/hashicorp/consul/blob/ba2f44ea0de39551c37728f995832c0b4a2ba2e2/agent/proxycfg/state.go#L344-L349)